File: Easy/[19] Remove Nth Node From End of List/Python3 Solution/[19] Remove Nth Node From End of List.py
"""
Accepted
19 [Easy]
Runtime: 39 ms, faster than 44.21% of Python3 online submissions for Remove Nth Node From End of List.
Memory Usage: 16.51 MB, less than 61.49% of Python3 online submissions for Remove Nth Node From End of List.
"""
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        count = 0
        ptr = head
        while ptr:
            count += 1
            ptr = ptr.next
        
        pn = count - n
        ptr, cur = head, head
        for i in range(pn):
            if i > 0:
                cur = cur.next
            ptr = ptr.next

        if ptr == head:
            head = head.next
            ptr = ptr.next
        else:
            cur.next = ptr.next
        
        return head

File: Easy/[19] Remove Nth Node From End of List/Java Solution/[19] Remove Nth Node From End of List.java
/*
Accepted
19 [Easy]
Runtime: 0 ms, faster than 100.00% of Java online submissions for Remove Nth Node From End of List.
Memory Usage: 40.63 MB, less than 99.96% of Java online submissions for Remove Nth Node From End of List.
 */
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode ptr=head,cur=head;
        int count=0,i;
        while(ptr!=null) {
            count++;
            ptr=ptr.next;
        }
        ptr=head;
        int pn=count-n;
        for(i=0;i<pn;i++) {
            if(i>0)
                cur=cur.next;
            ptr=ptr.next;
        }
        if(ptr==head) {
            head=ptr.next;
            ptr.next=null;
        }
        else
            cur.next=ptr.next;
        return head;
    }
}